src/Shared/Components/PrerenderComponentApplicationStore.cs

```diff
@@ -23,33 +30,103 @@ public PrerenderComponentApplicationStore(string existingState)
 
-        public Task<IDictionary<string, byte[]>> GetPersistedStateAsync()
+        public Task<IDictionary<string, ReadOnlySequence<byte>>> GetPersistedStateAsync()
         {
-            return Task.FromResult((IDictionary<string, byte[]>)ExistingState);
+            return Task.FromResult((IDictionary<string, ReadOnlySequence<byte>>)ExistingState);
         }
 
-        protected virtual byte[] SerializeState(IReadOnlyDictionary<string, byte[]> state)
+        protected virtual PooledByteBufferWriter SerializeState(IReadOnlyDictionary<string, ReadOnlySequence<byte>> state)
         {
-            return JsonSerializer.SerializeToUtf8Bytes(state);
+            // System.Text.Json doesn't support serializing ReadonlySequence<byte> so we need to buffer
+            // the data with a memory pool here. We will change our serialization strategy in the future here
+            // so that we can avoid this step.
+            var buffer = new PooledByteBufferWriter();
+            var jsonWriter = new Utf8JsonWriter(buffer);
+            jsonWriter.WriteStartObject();
+            foreach (var (key, value) in state)
+            {
+                if (value.IsSingleSegment)
+                {
+                    jsonWriter.WriteBase64String(key, value.First.Span);
+                }
+                else
+                {
+                    WriteMultipleSegments(jsonWriter, key, value);
+                }
+                jsonWriter.Flush();
+            }
+
+            jsonWriter.WriteEndObject();
+            jsonWriter.Flush();
+            return buffer;
+
+            static void WriteMultipleSegments(Utf8JsonWriter jsonWriter, string key, ReadOnlySequence<byte> value)
+            {
+                byte[] unescapedArray = null;
+                var valueLenght = (int)value.Length;
+
+                Span<byte> valueSegment = value.Length <= 256 ?
+                    stackalloc byte[valueLenght] :
+                    (unescapedArray = ArrayPool<byte>.Shared.Rent(valueLenght)).AsSpan().Slice(0, valueLenght);
+
+                value.CopyTo(valueSegment);
+                jsonWriter.WriteBase64String(key, valueSegment);
+
+                if (unescapedArray != null)
+                {
+                    valueSegment.Clear();
+                    ArrayPool<byte>.Shared.Return(unescapedArray);
+                }
+            }
         }
 
-        public Task PersistStateAsync(IReadOnlyDictionary<string, byte[]> state)
+        public Task PersistStateAsync(IReadOnlyDictionary<string, ReadOnlySequence<byte>> state)
         {
-            var bytes = SerializeState(state);
+            using var bytes = SerializeState(state);
+            var length = Base64.GetMaxEncodedToUtf8Length(bytes.WrittenCount);
+            // We can do this because the representation in bytes for characters in the base64 alphabet for utf-8 is 1.
+            _buffer = ArrayPool<char>.Shared.Rent(length);
```

Comment:
Can this function be called more than once? Should we return this buffer if it's non-null in that case?

Reply:
This is only called by the framework once at the right time during persistence, so we don't have to worry about incorrect usages.